src/stream/src/common/log_store_impl/only_writer_log_store.rs

```diff
@@ -0,0 +1,137 @@
+    _table_id: TableId,
+
+    seq_id: SeqIdType,
+
+    state_store: LS,
+
+    serde: LogStoreRowSerde,
+
+    _identity: String,
+
+    truncate_offset: Option<ReaderTruncationOffsetType>,
+}
+
+impl<LS: LocalStateStore> OnlyWriterLogStoreWriter<LS> {
+    pub(crate) fn new(
+        table_id: TableId,
+        state_store: LS,
+        serde: LogStoreRowSerde,
+        identity: String,
+    ) -> Self {
+        Self {
+            _table_id: table_id,
+            seq_id: FIRST_SEQ_ID,
+            state_store,
+            serde,
+            _identity: identity,
+            truncate_offset: None,
+        }
+    }
+
+    pub async fn init(
+        &mut self,
+        epoch: risingwave_common::util::epoch::EpochPair,
+        _pause_read_on_bootstrap: bool,
+    ) -> LogStoreResult<()> {
+        self.state_store
+            .init(InitOptions::new_with_epoch(epoch))
+            .await?;
+        self.seq_id = FIRST_SEQ_ID;
+        Ok(())
+    }
+
+    pub fn write_chunk(&mut self, chunk: StreamChunk) -> LogStoreResult<()> {
+        if chunk.cardinality() == 0 {
+            return Ok(());
+        }
+        let epoch = self.state_store.epoch();
+        let start_seq_id = self.seq_id;
+        self.seq_id += chunk.cardinality() as SeqIdType;
+        for (i, (op, row)) in chunk.rows().enumerate() {
+            let seq_id = start_seq_id + (i as SeqIdType);
+            let (_vnode, key, value) = self.serde.serialize_data_row(epoch, seq_id, op, row);
+            self.state_store.insert(key, value, None)?;
+        }
+        Ok(())
+    }
+
+    pub async fn flush_current_epoch(
+        &mut self,
+        next_epoch: u64,
+        is_checkpoint: bool,
+    ) -> LogStoreResult<()> {
+        let epoch = self.state_store.epoch();
+        for vnode in self.serde.vnodes().iter_vnodes() {
+            let (key, value) = self.serde.serialize_barrier(epoch, vnode, is_checkpoint);
+            self.state_store.insert(key, value, None)?;
+        }
+
+        let mut watermark = None;
+        if let Some(truncate_offset) = self.truncate_offset {
+            watermark = Some(VnodeWatermark::new(
+                self.serde.vnodes().clone(),
+                self.serde
+                    .serialize_truncation_offset_watermark(truncate_offset),
+            ));
+            self.truncate_offset = None;
+        }
+        self.state_store.flush(vec![]).await?;
+        let watermark = watermark.into_iter().collect_vec();
+        self.state_store.seal_current_epoch(
+            next_epoch,
+            SealCurrentEpochOptions::new(watermark, WatermarkDirection::Ascending),
+        );
+        self.seq_id = FIRST_SEQ_ID;
+        Ok(())
+    }
+
+    pub fn update_vnode_bitmap(&mut self, new_vnodes: Arc<Bitmap>) -> LogStoreResult<()> {
+        self.serde.update_vnode_bitmap(new_vnodes.clone());
+        Ok(())
+    }
+
+    pub fn truncate(&mut self, offset: TruncateOffset) -> LogStoreResult<()> {
```

Comment:
If the timeout stream is unnecessary, this truncate method and the truncate_offset field are also unnecessary. We can pass the truncate offset in `seal_current_epoch` directly.